components/usage/pkg/apiv1/usage.go

```diff
@@ -88,6 +97,212 @@ func (us *UsageService) ListBilledUsage(ctx context.Context, in *v1.ListBilledUs
+	from := req.GetStartTime().AsTime()
+	to := req.GetEndTime().AsTime()
+
+	if to.Before(from) {
+		return nil, status.Errorf(codes.InvalidArgument, "End time must be after start time")
+	}
+
+	report, err := s.reportGenerator.GenerateUsageReport(ctx, from, to)
+	if err != nil {
+		log.Log.WithError(err).Error("Failed to reconcile time range.")
+		return nil, status.Error(codes.Internal, "failed to reconcile time range")
+	}
+
+	err = db.CreateUsageRecords(ctx, s.conn, report.UsageRecords)
+	if err != nil {
+		log.Log.WithError(err).Error("Failed to persist usage records.")
+		return nil, status.Error(codes.Internal, "failed to persist usage records")
+	}
+
+	filename := fmt.Sprintf("%s.gz", time.Now().Format(time.RFC3339))
+	err = s.contentService.UploadUsageReport(ctx, filename, report.UsageRecords)
+	if err != nil {
+		log.Log.WithError(err).Error("Failed to persist usage report to content service.")
+		return nil, status.Error(codes.Internal, "failed to persist usage report to content service")
+	}
+
+	var sessions []*v1.BilledSession
+	for _, instance := range report.UsageRecords {
+		sessions = append(sessions, usageRecordToBilledUsageProto(instance))
+	}
+
+	return &v1.ReconcileUsageResponse{
+		Sessions: sessions,
+	}, nil
+
+}
+
+func NewUsageService(conn *gorm.DB, reportGenerator *ReportGenerator, contentSvc contentservice.Interface) *UsageService {
+	return &UsageService{
+		conn:            conn,
+		reportGenerator: reportGenerator,
+		contentService:  contentSvc,
+	}
+}
+
+func usageRecordToBilledUsageProto(usageRecord db.WorkspaceInstanceUsage) *v1.BilledSession {
+	var endTime *timestamppb.Timestamp
+	if usageRecord.StoppedAt.Valid {
+		endTime = timestamppb.New(usageRecord.StoppedAt.Time)
+	}
+	return &v1.BilledSession{
+		AttributionId:  string(usageRecord.AttributionID),
+		UserId:         usageRecord.UserID.String(),
+		WorkspaceId:    usageRecord.WorkspaceID,
+		TeamId:         "",
+		WorkspaceType:  string(usageRecord.WorkspaceType),
+		ProjectId:      usageRecord.ProjectID,
+		InstanceId:     usageRecord.InstanceID.String(),
+		WorkspaceClass: usageRecord.WorkspaceClass,
+		StartTime:      timestamppb.New(usageRecord.StartedAt),
+		EndTime:        endTime,
+		Credits:        usageRecord.CreditsUsed,
+	}
+}
+
+type InvalidSession struct {
+	Reason  string
+	Session db.WorkspaceInstanceForUsage
+}
+
+type UsageReport struct {
+	GenerationTime time.Time
+
+	From time.Time
+	To   time.Time
+
+	RawSessions     []db.WorkspaceInstanceForUsage
+	InvalidSessions []InvalidSession
+
+	UsageRecords []db.WorkspaceInstanceUsage
+}
+
+func NewReportGenerator(conn *gorm.DB, pricer *controller.WorkspacePricer) *ReportGenerator {
+	return &ReportGenerator{
+		conn:    conn,
+		pricer:  pricer,
+		nowFunc: time.Now,
+	}
+}
+
+type ReportGenerator struct {
+	conn    *gorm.DB
+	pricer  *controller.WorkspacePricer
+	nowFunc func() time.Time
+}
+
```

Comment:
Should we remove this `ReportGenerator` to it's own file now.? This `usage.go` is now rather large.

Reply:
Sure, that's easy enough :)

Reply:
Done and rebased. Also moved the pricer into `apiv1` as it didn't make sense in the controller anymore